File: packages/cli/src/commands/quick.ts
import { Command } from "commander";
import * as fs from "node:fs";
import * as path from "node:path";
import {
  type Lang,
  resolveLang,
  quickTitle,
  quickRootLabel,
  quickRecommendationLabel,
  quickReasonLabel,
  quickDetectedSignalsLabel,
  quickNoneLabel,
  quickNextStepsLabel,
} from "../i18n.js";

interface QuickResult {
  root: string;
  recommendation: string;
  reason: string;
  next_steps: string[];
  detected_signals: string[];
}

const harnessMarkers = ["AGENTS.md", "X_HARNESS.md", ".x-harness"];
const cardNames = [
  "completion-card.yaml",
  "completion-card.yml",
  "completion-card.json",
];

function detectSignals(root: string): string[] {
  const signals: string[] = [];
  for (const marker of harnessMarkers) {
    const p = path.join(root, marker);
    if (fs.existsSync(p)) {
      signals.push(`harness_marker:${marker}`);
    }
  }
  // Walk up to depth 4 looking for completion cards
  function walk(dir: string, depth: number) {
    if (depth > 4) return;
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (entry.isDirectory()) {
        const name = entry.name;
        if (
          name === "node_modules" ||
          name === ".git" ||
          name === "vendor" ||
          name === "dist" ||
          name === "coverage"
        ) {
          continue;
        }
        // Skip generated harness state directories
        if (name === "tmp" || name === "cache") {
          if (path.basename(dir) === ".x-harness") {
            continue;
          }
        }
        walk(path.join(dir, name), depth + 1);
      } else {
        for (const name of cardNames) {
          if (entry.name === name) {
            const rel = path.relative(root, path.join(dir, entry.name));
            signals.push(`completion_card:${rel}`);
            break;
          }
        }
      }
    }
  }
  walk(root, 0);
  return signals;
}

function buildRecommendation(
  root: string,
  signals: string[]
): { recommendation: string; reason: string; nextSteps: string[] } {
  let hasHarness = false;
  const cardPaths: string[] = [];
  for (const s of signals) {
    if (s.startsWith("harness_marker:")) {
      hasHarness = true;
    }
    if (s.startsWith("completion_card:")) {
      cardPaths.push(s.slice("completion_card:".length));
    }
  }

  let recommendation: string;
  let reason: string;
  const nextSteps: string[] = [];

  if (!hasHarness) {
    recommendation = "xh start";
    reason =
      "No harness markers found under root. Begin with guided onboarding.";
    nextSteps.push("xh start");
    nextSteps.push("xh init");
  } else if (cardPaths.length > 0) {
    recommendation = `xh check --card ${cardPaths[0]}`;
    reason = "A completion card was found. Verify it as the next step.";
    nextSteps.push(`xh check --card ${cardPaths[0]}`);
  } else {
    recommendation = `xh doctor --root ${root} --json`;
    reason =
      "Harness is present but no completion card found yet. Check workspace health first.";
    nextSteps.push(`xh doctor --root ${root} --json`);
  }

  nextSteps.push("xh run builtin:ci --dry-run");
  nextSteps.push("xh learn");

  return { recommendation, reason, nextSteps };
}

export function quickCommand(): Command {
  return new Command("quick")
    .description("Read-only next-action recommender for newcomers")
    .option("--root <path>", "Repository root", process.cwd())
    .option("--json", "Output JSON instead of text", false)
    .option("--lang <code>", "Language", "en")
    .action(
      async (
        opts: { root: string; json: boolean; lang: string },
        cmd: Command
      ) => {
        const lang: Lang = resolveLang(opts, cmd.parent?.opts() ?? {});
        const root = path.resolve(opts.root);
        const signals = detectSignals(root);
        const { recommendation, reason, nextSteps } = buildRecommendation(
          root,
          signals
        );

        const result: QuickResult = {
          root,
          recommendation,
          reason,
          next_steps: nextSteps,
          detected_signals: signals,
        };

        if (opts.json) {
          console.log(JSON.stringify(result, null, 2));
        } else {
          console.log(`# ${quickTitle(lang)}`);
          console.log("");
          console.log(`${quickRootLabel(lang)}: ${result.root}`);
          console.log(
            `${quickRecommendationLabel(lang)}: ${result.recommendation}`
          );
          console.log(`${quickReasonLabel(lang)}: ${result.reason}`);
          console.log("");
          console.log(quickDetectedSignalsLabel(lang));
          if (result.detected_signals.length === 0) {
            console.log(quickNoneLabel(lang));
          } else {
            for (const s of result.detected_signals) {
              console.log(`  - ${s}`);
            }
          }
          console.log("");
          console.log(quickNextStepsLabel(lang));
          for (const s of result.next_steps) {
            console.log(`  - ${s}`);
          }
        }
      }
    );
}

File: internal/cli/quick.go
package cli

import (
	"fmt"
	"io"
	"os"
	"path/filepath"

	"github.com/BrianNguyen29/x-harness/internal/repo"
)

// QuickResult is the JSON output shape for the quick command.
type QuickResult struct {
	Root            string   `json:"root"`
	Recommendation  string   `json:"recommendation"`
	Reason          string   `json:"reason"`
	NextSteps       []string `json:"next_steps"`
	DetectedSignals []string `json:"detected_signals"`
}

func handleQuick(args []string, stdout io.Writer, stderr io.Writer, lang Lang) int {
	root := ""
	jsonMode := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "-h", "--help":
			fmt.Fprintln(stderr, "usage: xh quick [--root <path>] [--json]")
			return ExitUsage
		case "--root":
			if i+1 < len(args) {
				root = args[i+1]
				i++
			}
		case "--json":
			jsonMode = true
		}
	}

	// Resolve root
	resolvedRoot := root
	if resolvedRoot == "" {
		found, err := repo.FindRoot("")
		if err == nil {
			resolvedRoot = found
		} else {
			wd, err := os.Getwd()
			if err != nil {
				fmt.Fprintf(stderr, "error: %v\n", err)
				return ExitError
			}
			resolvedRoot = wd
		}
	}
	resolvedRoot, _ = filepath.Abs(resolvedRoot)

	signals := detectSignals(resolvedRoot)
	if signals == nil {
		signals = []string{}
	}
	recommendation, reason, nextSteps := buildRecommendation(resolvedRoot, signals)

	result := QuickResult{
		Root:            resolvedRoot,
		Recommendation:  recommendation,
		Reason:          reason,
		NextSteps:       nextSteps,
		DetectedSignals: signals,
	}

	if jsonMode {
		if err := WriteJSON(stdout, result); err != nil {
			return ExitError
		}
	} else {
		WriteLine(stdout, "# %s", quickTitle(lang))
		WriteLine(stdout, "")
		WriteLine(stdout, "%s: %s", quickRootLabel(lang), result.Root)
		WriteLine(stdout, "%s: %s", quickRecommendationLabel(lang), result.Recommendation)
		WriteLine(stdout, "%s: %s", quickReasonLabel(lang), result.Reason)
		WriteLine(stdout, "")
		WriteLine(stdout, "%s", quickDetectedSignalsLabel(lang))
		if len(result.DetectedSignals) == 0 {
			WriteLine(stdout, "%s", quickNoneLabel(lang))
		} else {
			for _, s := range result.DetectedSignals {
				WriteLine(stdout, "  - %s", s)
			}
		}
		WriteLine(stdout, "")
		WriteLine(stdout, "%s", quickNextStepsLabel(lang))
		for _, s := range result.NextSteps {
			WriteLine(stdout, "  - %s", s)
		}
	}

	return ExitOK
}

// harnessMarkers are files/directories that indicate a harness is present.
var harnessMarkers = []string{
	"AGENTS.md",
	"X_HARNESS.md",
	".x-harness",
}

// cardNames are common completion card filenames to look for.
var cardNames = []string{
	"completion-card.yaml",
	"completion-card.yml",
	"completion-card.json",
}

func detectSignals(root string) []string {
	var signals []string
	for _, marker := range harnessMarkers {
		path := filepath.Join(root, marker)
		if _, err := os.Stat(path); err == nil {
			signals = append(signals, "harness_marker:"+marker)
		}
	}
	// Look for completion cards recursively under root, limited depth for speed.
	_ = filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			name := d.Name()
			// Skip deep or irrelevant directories to keep it fast
			if name == "node_modules" || name == ".git" || name == "vendor" || name == "dist" || name == "coverage" {
				return filepath.SkipDir
			}
			// Skip generated harness state directories
			if name == "tmp" || name == "cache" {
				parent := filepath.Dir(path)
				if filepath.Base(parent) == ".x-harness" {
					return filepath.SkipDir
				}
			}
			// Limit depth by skipping very deep walks: if we're more than 4 levels
			// below root, skip further descent. This is a simple heuristic.
			rel, _ := filepath.Rel(root, path)
			if rel != "" && rel != "." {
				depth := 0
				for i := 0; i < len(rel); i++ {
					if rel[i] == filepath.Separator {
						depth++
					}
				}
				if depth >= 4 {
					return filepath.SkipDir
				}
			}
			return nil
		}
		for _, name := range cardNames {
			if d.Name() == name {
				rel, _ := filepath.Rel(root, path)
				signals = append(signals, "completion_card:"+rel)
				return nil
			}
		}
		return nil
	})
	return signals
}

func buildRecommendation(root string, signals []string) (recommendation, reason string, nextSteps []string) {
	hasHarness := false
	var cardPaths []string
	for _, s := range signals {
		if len(s) > 15 && s[:15] == "harness_marker:" {
			hasHarness = true
		}
		if len(s) > 16 && s[:16] == "completion_card:" {
			cardPaths = append(cardPaths, s[16:])
		}
	}

	if !hasHarness {
		recommendation = "xh start"
		reason = "No harness markers found under root. Begin with guided onboarding."
		nextSteps = append(nextSteps, "xh start")
		nextSteps = append(nextSteps, "xh init")
	} else if len(cardPaths) > 0 {
		recommendation = "xh check --card " + cardPaths[0]
		reason = "A completion card was found. Verify it as the next step."
		nextSteps = append(nextSteps, "xh check --card "+cardPaths[0])
	} else {
		recommendation = "xh doctor --root " + root + " --json"
		reason = "Harness is present but no completion card found yet. Check workspace health first."
		nextSteps = append(nextSteps, "xh doctor --root "+root+" --json")
	}

	// Always include these safe, read-only next steps
	nextSteps = append(nextSteps, "xh run builtin:ci --dry-run")
	nextSteps = append(nextSteps, "xh learn")

	return recommendation, reason, nextSteps
}
